libs/test_positivity.py

```diff
@@ -142,7 +153,32 @@ def run(
 
         return AllMethods(all_methods_timeseries=all_wide, test_positivity=test_positivity)
 
+    @staticmethod
+    def _methods_with_columns_available(
+        methods_in: Sequence[Method], available_columns: Sequence[str]
+    ) -> Sequence[Method]:
+        available_columns_set = set(available_columns)
+        return [m for m in methods_in if m.columns <= available_columns_set]
+
+    @staticmethod
+    def _list_columns(methods: Iterable[Method]) -> List[FieldName]:
+        """Returns unsorted list of columns in given Method objects."""
+        return list(set(chain.from_iterable(method.columns for method in methods)))
+
     def write(self, csv_path: pathlib.Path):
         self.all_methods_timeseries.to_csv(
             csv_path, date_format="%Y-%m-%d", index=True, float_format="%.05g",
         )
+
+
+def run_and_maybe_join_columns(
+    mrts: timeseries.MultiRegionTimeseriesDataset, log
+) -> timeseries.MultiRegionTimeseriesDataset:
+    """Calculates test positivity and joins it with the input if it finishes"""
+    try:
+        test_positivity_results = AllMethods.run(mrts)
+    except TestPositivityException:
+        log.exception("test_positivity failed")
+        return mrts
```

Comment:
Yes. https://softwareengineering.stackexchange.com/a/351121
Though in this case I'm not expecting TestPositivityException to be raised in normal production runs.

Reply:
🤯 

At least this is well-contained to throw/catch the exceptions within a single class so not too problematic from an understandability perspective.  Would it be worth making run() private?  Else, should it have a docstring or whatever that documents the exceptions it can throw?

Reply:
I expect the calling code to not handle the different subtypes of TestPositivityException differently. The subtypes exist to help someone debugging what broke or why test positivity isn't working. Added a little doc string. run is public because the little cli/api.py generate_test_positivity calls it.